File: src/radio/WinLinkMail.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Globalization;

namespace HTCommander.radio
{
    public class WinLinkMail
    {
        public string MID { get; set; }
        public DateTime DateTime { get; set; }
        public string From { get; set; }
        public string To { get; set; }
        public string Subject { get; set; }
        public string Mbo { get; set; }
        public string Body { get; set; }
        public string Tag { get; set; }
        public string Location { get; set; }
        public string Attachements { get; set; }
        public int Flags { get; set; } // 1 = Unread
        public int Mailbox { get; set; }

        public enum MailFlags: int
        {
            Unread = 1,
            Private = 2,
            P2P = 4
        }

        public WinLinkMail() { }

        /*
        public static void Test()
        {
            string test1 = "MID: 1CCIZGEQKFAC\r\nDate: 2025/02/22 03:30\r\nType: Private\r\nFrom: KK7VZT\r\nTo: ysainthilaire@hotmail.com\r\nSubject: Test2\r\nMbo: KK7VZT\r\nX-Location: 45.395833N, 122.791667W (Grid square)\r\nBody: 5\r\n\r\ntest2";
            string test2 = "MID: OLRJZ16F3KHG\r\nDate: 2025/02/22 19:09\r\nType: Private\r\nFrom: KK7VZT\r\nTo: ysainthilaire@hotmail.com\r\nSubject: Test4\r\nMbo: KK7VZT\r\nX-Location: 45.395833N, 122.791667W (Grid square)\r\nBody: 214\r\n\r\nThis is a sample message.\r\nThis is a sample message.\r\nThis is a sample message.\r\nThis is a sample message.\r\nThis is a sample message.\r\nThis is a sample message.\r\nThis is a sample message.\r\nThis is a sample message.";
            WinLinkMail x1 = WinLinkMail.DeserializeMail(test1);
            WinLinkMail x2 = WinLinkMail.DeserializeMail(test2);
            string m1 = SerializeMail(x1);
            string m2 = SerializeMail(x2);

            List<byte[]> blocks = EncodeMailToBlocks(x1);
            WinLinkMail x3 = DecodeBlocksToEmail(blocks);
        }

        public static bool Test2()
        {
            string xm1 = "8A34C7000000ECF57A1C6D66F79F7F89E6E9F47BBD7E9736D6672D87ED00F8E160EFB7961C1DDD7D2A3AD354A1BFA14D52D6D3C00BFCA805FB9FEFA81500825CCB99EFDFE6955BA77C3F15F51C50E4BB8E517FECE77F565F46BF86D198D8F322DCB49688BC56EBDF096CD99DF01F77D993EC16DB62F23CE6914315EA40BF0E3BF26E7B06282D35CE8E6D9E0574026E297E2321BB5B86B0155CB49B091E10E90F187697B0D25C047355ECDFE06D4E379C8A6126C0C4E3503CEE1122";
            byte[] m1 = Utils.HexStringToByteArray(xm1);
            byte[] d1 = new byte[199];
            int dlen1 = WinlinkCompression.Decode(m1, ref d1, true, 199);
            string ds1 = UTF8Encoding.UTF8.GetString(d1);
            WinLinkMail mm1 = WinLinkMail.DeserializeMail(ds1);
            string ds2 = WinLinkMail.SerializeMail(mm1);
            byte[] re1 = new byte[0];
            int clen1 = WinlinkCompression.Encode(UTF8Encoding.UTF8.GetBytes(ds2), ref re1, true);
            string rm1 = Utils.BytesToHex(re1);
            return (xm1 == rm1);
        }
        */

        public static WinLinkMail DecodeBlocksToEmail(List<byte[]> blocks)
        {
            // TODO: Add support for multiple blocks
            if (blocks == null) return null;
            if (blocks.Count == 0) return null;

            byte[] block = blocks[0];

            int cmdlen, ptr = 0;
            byte[] payload = null;
            while (ptr < block.Length)
            {
                int cmd = block[ptr];
                switch (cmd)
                {
                    case 1:
                        cmdlen = block[ptr + 1];
                        ptr += (2 + cmdlen);
                        break;
                    case 2:
                        cmdlen = block[ptr + 1];
                        payload = new byte[cmdlen];
                        Array.Copy(block, ptr + 2, payload, 0, cmdlen);
                        ptr += (2 + cmdlen);
                        break;
                    case 4:
                        cmdlen = block[ptr + 1];
                        if (WinLinkChecksum.ComputeChecksum(payload) != cmdlen) return null;
                        ptr += 2;
                        break;
                }
                
            }

            // Decompress the mail
            byte[] obuf = null;
            int expectedLength = (payload[2] + (payload[3] << 8) + (payload[4] << 16) + (payload[5] << 24));
            int obuflen = WinlinkCompression.Decode(payload, ref obuf, true, expectedLength);
            if (obuflen != expectedLength) return null;

            // Decode the mail
            return WinLinkMail.DeserializeMail(UTF8Encoding.UTF8.GetString(obuf));
        }

        public static List<byte[]> EncodeMailToBlocks(WinLinkMail mail)
        {
            byte[] payloadBuf = null;
            WinlinkCompression.Encode(UTF8Encoding.UTF8.GetBytes(WinLinkMail.SerializeMail(mail)), ref payloadBuf, true);
            if (payloadBuf == null) return null;
            byte[] subjectBuf = UTF8Encoding.UTF8.GetBytes(mail.Subject);
            List<byte[]> blocks = new List<byte[]>();

            // TODO: This assumes the message fits in one block, add support for multi-blocks.
            int ptr = 0;
            byte[] output = new byte[2 + subjectBuf.Length + 5 + payloadBuf.Length + 2];
            output[ptr++] = 1;
            output[ptr++] = (byte)(subjectBuf.Length + 3);
            Array.Copy(subjectBuf, 0, output, ptr, subjectBuf.Length);
            ptr += subjectBuf.Length;
            output[ptr++] = 0;
            output[ptr++] = 0x30; // ASCII '0' in HEX.
            output[ptr++] = 0;
            output[ptr++] = 2;
            output[ptr++] = (byte)payloadBuf.Length;
            Array.Copy(payloadBuf, 0, output, ptr, payloadBuf.Length);
            ptr += payloadBuf.Length;
            output[ptr++] = 4;
            output[ptr++] = WinLinkChecksum.ComputeChecksum(payloadBuf);

            blocks.Add(output);
            return blocks;
        }

        public static string SerializeMail(WinLinkMail mail)
        {
            StringBuilder sb = new StringBuilder();
            sb.AppendLine($"MID: {mail.MID}");
            sb.AppendLine($"Date: {mail.DateTime.ToString("yyyy/MM/dd HH:mm")}");
            if ((mail.Flags & (int)MailFlags.Private) != 0) { sb.AppendLine($"Type: Private"); }
            if (!string.IsNullOrEmpty(mail.From)) { sb.AppendLine($"From: {mail.From}"); }
            if (!string.IsNullOrEmpty(mail.To)) { sb.AppendLine($"To: {mail.To}"); }
            if (!string.IsNullOrEmpty(mail.Subject)) { sb.AppendLine($"Subject: {mail.Subject}"); }
            if (!string.IsNullOrEmpty(mail.Mbo)) { sb.AppendLine($"Mbo: {mail.Mbo}"); }
            if ((mail.Flags & (int)MailFlags.P2P) != 0) { sb.AppendLine($"X-P2P: True"); }
            if (!string.IsNullOrEmpty(mail.Location)) { sb.AppendLine($"X-Location: {mail.Location}"); }
            if (!string.IsNullOrEmpty(mail.Body)) { sb.AppendLine($"Body: " + mail.Body.Length); sb.AppendLine(); sb.Append(mail.Body); }
            return sb.ToString();
        }

        // https://winlink.org/sites/default/files/downloads/winlink_data_flow_and_data_packaging.pdf
        public static WinLinkMail DeserializeMail(string data)
        {
            WinLinkMail currentMail = new WinLinkMail();

            bool done = false;
            int i, bodyLength = -1;
            string[] lines = data.Replace("\r\n","\n").Split(new[] { '\n', '\r' });
            foreach (string line in lines)
            {
                if (done) continue;
                i = line.IndexOf(':');
                if (i > 0)
                {
                    string key = line.Substring(0, i).ToLower().Trim();
                    string value = line.Substring(i + 1).Trim();
            
                    switch (key)
                    {
                        case "": done = true; break;
                        case "mid": currentMail.MID = value; break;
                        case "date": currentMail.DateTime = DateTime.ParseExact(value, "yyyy/MM/dd HH:mm", CultureInfo.InvariantCulture); break;
                        case "type": { if (value.ToLower() == "private") { currentMail.Flags |= (int)MailFlags.Private; }; } break;
                        case "to": currentMail.To = value; break;
                        case "from": currentMail.From = value; break;
                        case "subject": currentMail.Subject = value; break;
                        case "mbo": currentMail.Mbo = value; break;
                        case "body": bodyLength = int.Parse(value); break;
                        case "x-location": currentMail.Location = value; break;
                        case "x-p2p": { if (value.ToLower() == "true") { currentMail.Flags |= (int)MailFlags.P2P; }; } break;
                    }
                }
            }

            i = data.IndexOf("\r\n\r\n");
            if (i > 0) { currentMail.Body = data.Substring(i + 4, bodyLength); }
            return currentMail;
        }

        // Serialize a list of stations to a plain text format
        public static string Serialize(List<WinLinkMail> mails)
        {
            StringBuilder sb = new StringBuilder();
            foreach (WinLinkMail mail in mails)
            {
                sb.AppendLine("Mail:");
                sb.AppendLine($"MID={mail.MID}");
                sb.AppendLine($"Time={mail.DateTime.ToString("o")}");
                if (!string.IsNullOrEmpty(mail.From)) { sb.AppendLine($"From={mail.From}"); }
                if (!string.IsNullOrEmpty(mail.To)) { sb.AppendLine($"To={mail.To}"); }
                sb.AppendLine($"Subject={mail.Subject}");
                if (!string.IsNullOrEmpty(mail.Mbo)) { sb.AppendLine($"Mbo={mail.Mbo}"); }
                sb.AppendLine($"Body={EscapeString(mail.Body)}");
                if (!string.IsNullOrEmpty(mail.Tag)) { sb.AppendLine($"Tag={mail.Tag}"); }
                if (!string.IsNullOrEmpty(mail.Location)) { sb.AppendLine($"Tag={mail.Location}"); }
                if (mail.Flags != 0) { sb.AppendLine($"Flags={(int)mail.Flags}"); }
                sb.AppendLine($"Mailbox={(int)mail.Mailbox}");
                sb.AppendLine(); // Separate entries with a blank line
            }
            return sb.ToString();
        }

        // Deserialize a plain text format into a list of StationInfoClass objects
        public static List<WinLinkMail> Deserialize(string data)
        {
            List<WinLinkMail> mails = new List<WinLinkMail>();
            WinLinkMail currentMail = null;

            string[] lines = data.Split(new[] { '\n', '\r' }, StringSplitOptions.RemoveEmptyEntries);
            foreach (string line in lines)
            {
                string trimmedLine = line.Trim();
                if (trimmedLine == "Mail:")
                {
                    if (currentMail != null) {
                        if (string.IsNullOrEmpty(currentMail.MID)) { currentMail.MID = WinLinkMail.GenerateMID(); }
                        mails.Add(currentMail);
                    }
                    currentMail = new WinLinkMail();
                }
                else if (currentMail != null)
                {
                    int i = trimmedLine.IndexOf('=');
                    if (i > 0)
                    {
                        string key = trimmedLine.Substring(0, i).Trim();
                        string value = trimmedLine.Substring(i + 1).Trim();

                        switch (key)
                        {
                            case "MID": currentMail.MID = value; break;
                            case "Time": currentMail.DateTime = DateTime.ParseExact(value, "o", System.Globalization.CultureInfo.InvariantCulture); break;
                            case "From": currentMail.From = value; break;
                            case "To": currentMail.To = value; break;
                            case "Subject": currentMail.Subject = value; break;
                            case "Mbo": currentMail.Mbo = value; break;
                            case "Body": currentMail.Body = UnescapeString(value); break;
                            case "Tag": currentMail.Tag = value; break;
                            case "Location": currentMail.Location = value; break;
                            case "Flags": currentMail.Flags = int.Parse(value); break;
                            case "Mailbox": currentMail.Mailbox = int.Parse(value); break;
                        }
                    }
                }
            }

            if (currentMail != null)
            {
                if (string.IsNullOrEmpty(currentMail.MID)) { currentMail.MID = WinLinkMail.GenerateMID(); }
                mails.Add(currentMail);
            }

            return mails;
        }


        private const char FieldSeparator = ';';
        private const char RecordSeparator = '\n';
        private const char EscapeCharacter = '\\';
        private static string EscapeString(string data)
        {
            if (string.IsNullOrEmpty(data)) return data;

            StringBuilder sb = new StringBuilder();
            foreach (char c in data)
            {
                if (c == FieldSeparator || c == RecordSeparator || c == EscapeCharacter)
                {
                    sb.Append(EscapeCharacter).Append(c);
                }
                else
                {
                    sb.Append(c);
                }
            }
            return sb.ToString();
        }

        private static string UnescapeString(string escapedData)
        {
            if (string.IsNullOrEmpty(escapedData)) return escapedData;

            StringBuilder sb = new StringBuilder();
            bool escaping = false;
            foreach (char c in escapedData)
            {
                if (escaping)
                {
                    sb.Append(c); // Append the escaped character directly
                    escaping = false;
                }
                else if (c == EscapeCharacter)
                {
                    escaping = true; // Next character is escaped
                }
                else
                {
                    sb.Append(c); // Normal character
                }
            }
            return sb.ToString();
        }

        public static string GenerateMID()
        {
            using (var rng = RandomNumberGenerator.Create())
            {
                var bytes = new byte[12];
                rng.GetBytes(bytes);

                StringBuilder result = new StringBuilder(12);
                foreach (byte b in bytes)
                {
                    // Map byte to alphanumeric characters (0-9, A-Z)
                    int value = b % 36; // 36 = 10 digits + 26 letters

                    if (value < 10)
                    {
                        // Digits 0-9
                        result.Append((char)('0' + value));
                    }
                    else
                    {
                        // Uppercase letters A-Z
                        result.Append((char)('A' + (value - 10)));
                    }
                }

                return result.ToString();
            }
        }
    }
}

File: lib/winlink/winlink_mail.dart
/*
Copyright 2026 Ylian Saint-Hilaire
Licensed under the Apache License, Version 2.0 (the "License");
http://www.apache.org/licenses/LICENSE-2.0
*/

import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'winlink_utils.dart';

/// A file attached to a [WinLinkMail].
class WinLinkMailAttachement {
  WinLinkMailAttachement({this.name = '', Uint8List? data})
    : data = data ?? Uint8List(0);

  String name;
  Uint8List data;
}

/// Bit flags applied to a [WinLinkMail].
enum MailFlags {
  unread(1),
  private(2),
  p2p(4);

  const MailFlags(this.value);
  final int value;
}

/// Result of [WinLinkMail.decodeBlocksToEmail].
class WinLinkMailDecodeResult {
  WinLinkMailDecodeResult(this.mail, this.fail, this.dataConsumed);

  final WinLinkMail? mail;
  final bool fail;
  final int dataConsumed;
}

/// Result of [WinLinkMail.encodeMailToBlocks].
class WinLinkMailEncodeResult {
  WinLinkMailEncodeResult(
    this.blocks,
    this.uncompressedSize,
    this.compressedSize,
  );

  final List<Uint8List>? blocks;
  final int uncompressedSize;
  final int compressedSize;
}

/// A Winlink email message.
class WinLinkMail {
  WinLinkMail();

  String? mid;
  DateTime dateTime = DateTime.now();
  String? from;
  String? to;
  String? cc;
  String? subject;
  String? mbo;
  String? body;
  String? tag;
  String? location;
  List<WinLinkMailAttachement>? attachments;
  int flags = 0; // 1 = Unread
  // Mailbox name (Inbox, Outbox, Draft, Sent, Archive, Trash, or custom)
  String mailbox = 'Inbox';

  /// Decodes a sequence of B2F blocks into a [WinLinkMail].
  static WinLinkMailDecodeResult decodeBlocksToEmail(Uint8List? block) {
    if (block == null || block.isEmpty) {
      return WinLinkMailDecodeResult(null, false, 0);
    }

    // Figure out if we have a full mail and the size of the mail
    int cmdlen;
    int payloadLen = 0;
    int ptr = 0;
    bool completeMail = false;
    while (!completeMail && ((ptr + 1) < block.length)) {
      final cmd = block[ptr];
      switch (cmd) {
        case 1:
          cmdlen = block[ptr + 1];
          ptr += (2 + cmdlen);
          break;
        case 2:
          cmdlen = block[ptr + 1];
          payloadLen += cmdlen;
          ptr += (2 + cmdlen);
          break;
        case 4:
          ptr += 2;
          completeMail = true;
          break;
        default:
          return WinLinkMailDecodeResult(null, false, 0);
      }
    }
    if (!completeMail) return WinLinkMailDecodeResult(null, false, 0);

    ptr = 0;
    final payload = Uint8List(payloadLen);
    int payloadPtr = 0;
    completeMail = false;
    while (!completeMail && ((ptr + 1) < block.length)) {
      final cmd = block[ptr];
      switch (cmd) {
        case 1:
          cmdlen = block[ptr + 1];
          ptr += (2 + cmdlen);
          break;
        case 2:
          cmdlen = block[ptr + 1];
          payload.setRange(payloadPtr, payloadPtr + cmdlen, block, ptr + 2);
          payloadPtr += cmdlen;
          ptr += (2 + cmdlen);
          break;
        case 4:
          cmdlen = block[ptr + 1];
          if (WinLinkChecksum.computeChecksum(payload) != cmdlen) {
            return WinLinkMailDecodeResult(null, true, 0);
          }
          ptr += 2;
          completeMail = true;
          break;
      }
    }

    // Decompress the mail
    final expectedLength =
        payload[2] +
        (payload[3] << 8) +
        (payload[4] << 16) +
        (payload[5] << 24);
    WinlinkDecodeResult? decoded;
    try {
      decoded = WinlinkCompression.decode(
        payload,
        checkCRC: true,
        expectedSize: expectedLength,
      );
    } catch (_) {
      decoded = null;
    }
    if (decoded == null || decoded.count != expectedLength) {
      return WinLinkMailDecodeResult(null, true, 0);
    }

    // Decode the mail
    final mail = deserializeMail(decoded.data);
    if (mail == null) return WinLinkMailDecodeResult(null, true, 0);
    return WinLinkMailDecodeResult(mail, false, ptr);
  }

  /// Encodes [mail] into a list of 128-byte B2F blocks.
  static WinLinkMailEncodeResult encodeMailToBlocks(WinLinkMail mail) {
    final uncompressedMail = serializeMail(mail);
    final uncompressedSize = uncompressedMail.length;
    final payloadBuf = WinlinkCompression.encode(
      uncompressedMail,
      prependCRC: true,
    );
    final subjectBuf = utf8.encode(mail.subject ?? '');
    final blocks = <Uint8List>[];

    // Encode the binary header
    final memoryStream = BytesBuilder();
    memoryStream.addByte(0x01);
    memoryStream.addByte((subjectBuf.length + 3) & 0xFF);
    memoryStream.add(subjectBuf);
    memoryStream.addByte(0x00);
    memoryStream.addByte(0x30); // ASCII '0' in HEX.
    memoryStream.addByte(0x00);

    int payloadPtr = 0;
    while (payloadPtr < payloadBuf.length) {
      final blockSize = min(250, payloadBuf.length - payloadPtr);
      memoryStream.addByte(0x02);
      memoryStream.addByte(blockSize & 0xFF);
      memoryStream.add(payloadBuf.sublist(payloadPtr, payloadPtr + blockSize));
      payloadPtr += blockSize;
    }

    memoryStream.addByte(0x04);
    memoryStream.addByte(WinLinkChecksum.computeChecksum(payloadBuf));

    final output = memoryStream.toBytes();
    final compressedSize = output.length;

    // Break the output into 128 byte blocks
    int outputPtr = 0;
    while (outputPtr < output.length) {
      final blockSize = min(128, output.length - outputPtr);
      blocks.add(
        Uint8List.sublistView(output, outputPtr, outputPtr + blockSize),
      );
      outputPtr += blockSize;
    }

    return WinLinkMailEncodeResult(blocks, uncompressedSize, compressedSize);
  }

  static Uint8List serializeMail(WinLinkMail mail) {
    final memoryStream = BytesBuilder();
    final bodyData = utf8.encode(mail.body ?? '');
    final between = Uint8List.fromList(const [0x0D, 0x0A]);
    final end = Uint8List.fromList(const [0x00]);

    final sb = StringBuffer();
    sb.write('MID: ${mail.mid}\r\n');
    sb.write('Date: ${_formatDate(mail.dateTime)}\r\n');
    if ((mail.flags & MailFlags.private.value) != 0) {
      sb.write('Type: Private\r\n');
    }
    if (_notEmpty(mail.from)) sb.write('From: ${mail.from}\r\n');
    if (_notEmpty(mail.to)) sb.write('To: ${mail.to}\r\n');
    if (_notEmpty(mail.cc)) sb.write('Cc: ${mail.cc}\r\n');
    if (_notEmpty(mail.subject)) sb.write('Subject: ${mail.subject}\r\n');
    if (_notEmpty(mail.mbo)) sb.write('Mbo: ${mail.mbo}\r\n');
    if ((mail.flags & MailFlags.p2p.value) != 0) sb.write('X-P2P: True\r\n');
    if (_notEmpty(mail.location)) sb.write('X-Location: ${mail.location}\r\n');
    if (_notEmpty(mail.body)) sb.write('Body: ${bodyData.length}\r\n');
    if (mail.attachments != null) {
      for (final attachement in mail.attachments!) {
        sb.write('File: ${attachement.data.length} ${attachement.name}\r\n');
      }
    }
    sb.write('\r\n');

    // Assemble the binary email
    memoryStream.add(utf8.encode(sb.toString()));
    memoryStream.add(bodyData);
    memoryStream.add(between);
    if (mail.attachments != null) {
      for (final attachement in mail.attachments!) {
        memoryStream.add(attachement.data);
        memoryStream.add(between);
      }
    }
    memoryStream.add(end);
    return memoryStream.toBytes();
  }

  static int findFirstDoubleNewline(Uint8List? data) {
    // Not enough data to contain \r\n\r\n
    if (data == null || data.length < 4) return -1;
    for (int i = 0; i <= data.length - 4; i++) {
      // Found \r\n\r\n at index i
      if (data[i] == 0x0D &&
          data[i + 1] == 0x0A &&
          data[i + 2] == 0x0D &&
          data[i + 3] == 0x0A) {
        return i;
      }
    }
    return -1; // \r\n\r\n not found
  }

  // https://winlink.org/sites/default/files/downloads/winlink_data_flow_and_data_packaging.pdf
  static WinLinkMail? deserializeMail(Uint8List databuf) {
    final currentMail = WinLinkMail();

    // Pull the header out of the data
    final headerLimit = findFirstDoubleNewline(databuf);
    if (headerLimit < 0) return null;
    final header = utf8.decode(
      databuf.sublist(0, headerLimit),
      allowMalformed: true,
    );

    // Decode the header
    bool done = false;
    int bodyLength = -1;
    int ptr = headerLimit + 4;
    final lines = header.replaceAll('\r\n', '\n').split(RegExp(r'[\n\r]'));
    for (final line in lines) {
      if (done) continue;
      final i = line.indexOf(':');
      if (i > 0) {
        final key = line.substring(0, i).toLowerCase().trim();
        final value = line.substring(i + 1).trim();

        switch (key) {
          case '':
            done = true;
            break;
          case 'mid':
            currentMail.mid = value;
            break;
          case 'date':
            currentMail.dateTime = _parseDate(value);
            break;
          case 'type':
            if (value.toLowerCase() == 'private') {
              currentMail.flags |= MailFlags.private.value;
            }
            break;
          case 'to':
            currentMail.to = value;
            break;
          case 'cc':
            currentMail.cc = value;
            break;
          case 'from':
            currentMail.from = value;
            break;
          case 'subject':
            currentMail.subject = value;
            break;
          case 'mbo':
            currentMail.mbo = value;
            break;
          case 'body':
            bodyLength = int.parse(value);
            break;
          case 'file':
            final j = value.indexOf(' ');
            if (j > 0) {
              final attachement = WinLinkMailAttachement(
                name: value.substring(j + 1).trim(),
                data: Uint8List(
                  int.parse(value.substring(0, j).toLowerCase().trim()),
                ),
              );
              currentMail.attachments ??= <WinLinkMailAttachement>[];
              currentMail.attachments!.add(attachement);
            }
            break;
          case 'x-location':
            currentMail.location = value;
            break;
          case 'x-p2p':
            if (value.toLowerCase() == 'true') {
              currentMail.flags |= MailFlags.p2p.value;
            }
            break;
        }
      }
    }

    // Pull the body out of the data
    if (bodyLength > 0) {
      currentMail.body = utf8.decode(
        databuf.sublist(ptr, ptr + bodyLength),
        allowMalformed: true,
      );
      ptr += (bodyLength + 2);
    }

    // Pull the attachments out of the data
    if (currentMail.attachments != null) {
      for (final attachment in currentMail.attachments!) {
        attachment.data.setRange(0, attachment.data.length, databuf, ptr);
        ptr += (attachment.data.length + 2);
      }
    }

    return currentMail;
  }

  // Serialize a list of mails to a plain text format
  static String serialize(List<WinLinkMail> mails) {
    final sb = StringBuffer();
    for (final mail in mails) {
      sb.writeln('Mail:');
      sb.writeln('MID=${mail.mid}');
      sb.writeln('Time=${mail.dateTime.toIso8601String()}');
      if (_notEmpty(mail.from)) sb.writeln('From=${mail.from}');
      if (_notEmpty(mail.to)) sb.writeln('To=${mail.to}');
      if (_notEmpty(mail.cc)) sb.writeln('Cc=${mail.cc}');
      sb.writeln('Subject=${mail.subject}');
      if (_notEmpty(mail.mbo)) sb.writeln('Mbo=${mail.mbo}');
      sb.writeln('Body=${_escapeString(mail.body)}');
      if (_notEmpty(mail.tag)) sb.writeln('Tag=${mail.tag}');
      if (_notEmpty(mail.location)) sb.writeln('Tag=${mail.location}');
      if (mail.flags != 0) sb.writeln('Flags=${mail.flags}');
      if (_notEmpty(mail.mailbox)) sb.writeln('Mailbox=${mail.mailbox}');
      if (mail.attachments != null) {
        for (final attachement in mail.attachments!) {
          sb.writeln('File=${attachement.name}');
          sb.writeln('FileData=${base64.encode(attachement.data)}');
        }
      }
      sb.writeln(); // Separate entries with a blank line
    }
    return sb.toString();
  }

  // Deserialize a plain text format into a list of WinLinkMail objects
  static List<WinLinkMail> deserialize(String data) {
    final mails = <WinLinkMail>[];
    WinLinkMail? currentMail;

    String? fileName;
    final lines = data
        .split(RegExp(r'[\n\r]'))
        .where((l) => l.isNotEmpty)
        .toList();
    for (final line in lines) {
      final trimmedLine = line.trim();
      if (trimmedLine == 'Mail:') {
        if (currentMail != null) {
          if (!_notEmpty(currentMail.mid)) currentMail.mid = generateMID();
          mails.add(currentMail);
        }
        currentMail = WinLinkMail();
      } else if (currentMail != null) {
        final i = trimmedLine.indexOf('=');
        if (i > 0) {
          final key = trimmedLine.substring(0, i).trim();
          final value = trimmedLine.substring(i + 1).trim();

          switch (key) {
            case 'MID':
              currentMail.mid = value;
              break;
            case 'Time':
              currentMail.dateTime = DateTime.parse(value);
              break;
            case 'From':
              currentMail.from = value;
              break;
            case 'To':
              currentMail.to = value;
              break;
            case 'Cc':
              currentMail.cc = value;
              break;
            case 'Subject':
              currentMail.subject = value;
              break;
            case 'Mbo':
              currentMail.mbo = value;
              break;
            case 'Body':
              currentMail.body = _unescapeString(value);
              break;
            case 'Tag':
              currentMail.tag = value;
              break;
            case 'Location':
              currentMail.location = value;
              break;
            case 'Flags':
              currentMail.flags = int.parse(value);
              break;
            case 'Mailbox':
              // Support both old integer format and new string format
              final mailboxIndex = int.tryParse(value);
              if (mailboxIndex != null) {
                // Convert old integer to string name
                const defaultMailboxes = [
                  'Inbox',
                  'Outbox',
                  'Draft',
                  'Sent',
                  'Archive',
                  'Trash',
                ];
                currentMail.mailbox =
                    (mailboxIndex >= 0 &&
                        mailboxIndex < defaultMailboxes.length)
                    ? defaultMailboxes[mailboxIndex]
                    : 'Inbox';
              } else {
                currentMail.mailbox = value;
              }
              break;
            case 'File':
              fileName = value;
              break;
            case 'FileData':
              if (_notEmpty(fileName)) {
                currentMail.attachments ??= <WinLinkMailAttachement>[];
                final attachement = WinLinkMailAttachement(
                  name: fileName!,
                  data: base64.decode(value),
                );
                currentMail.attachments!.add(attachement);
                fileName = null;
              }
              break;
          }
        }
      }
    }

    if (currentMail != null) {
      if (!_notEmpty(currentMail.mid)) currentMail.mid = generateMID();
      mails.add(currentMail);
    }

    return mails;
  }

  static const String _fieldSeparator = ';';
  static const String _recordSeparator = '\n';
  static const String _escapeCharacter = '\\';

  static String? _escapeString(String? data) {
    if (data == null || data.isEmpty) return data;

    final sb = StringBuffer();
    for (final c in data.split('')) {
      if (c == _fieldSeparator ||
          c == _recordSeparator ||
          c == _escapeCharacter) {
        sb.write(_escapeCharacter);
        sb.write(c);
      } else {
        sb.write(c);
      }
    }
    return sb.toString();
  }

  static String? _unescapeString(String? escapedData) {
    if (escapedData == null || escapedData.isEmpty) return escapedData;

    final sb = StringBuffer();
    bool escaping = false;
    for (final c in escapedData.split('')) {
      if (escaping) {
        sb.write(c); // Append the escaped character directly
        escaping = false;
      } else if (c == _escapeCharacter) {
        escaping = true; // Next character is escaped
      } else {
        sb.write(c); // Normal character
      }
    }
    return sb.toString();
  }

  static String generateMID() {
    final rng = Random.secure();
    final result = StringBuffer();
    for (int i = 0; i < 12; i++) {
      // Map byte to alphanumeric characters (0-9, A-Z)
      final value = rng.nextInt(256) % 36; // 36 = 10 digits + 26 letters
      if (value < 10) {
        result.writeCharCode(0x30 + value); // Digits 0-9
      } else {
        result.writeCharCode(0x41 + (value - 10)); // Uppercase letters A-Z
      }
    }
    return result.toString();
  }

  static bool isMailForStation(
    String? callsign,
    String? to,
    String? cc, {
    required void Function(bool others) onOthers,
  }) {
    final r1 = _isMailForStationEx(callsign, to);
    final r2 = _isMailForStationEx(callsign, cc);
    onOthers(r1.others || r2.others);
    return r1.match || r2.match;
  }

  static ({bool match, bool others}) _isMailForStationEx(
    String? callsign,
    String? t,
  ) {
    bool others = false;
    bool response = false;
    if (callsign == null || callsign.isEmpty || t == null || t.isEmpty) {
      return (match: false, others: false);
    }
    final s = t.split(';');
    for (final s2 in s) {
      if (s2.isEmpty) continue;
      bool match = false;
      final i = s2.indexOf('@');
      if (i == -1) {
        // Callsign
        if (callsign.toUpperCase() == s2.toUpperCase()) match = true;
        if (s2.toUpperCase().startsWith('${callsign.toUpperCase()}-')) {
          match = true;
        }
      } else {
        // Email
        final key = s2.substring(0, i).toUpperCase();
        final value = s2.substring(i + 1).toUpperCase();
        if ((value == 'WINLINK.ORG' && callsign.toUpperCase() == key) ||
            (key.toUpperCase().startsWith('$callsign-'))) {
          match = true;
        }
      }
      if (match) {
        response = true;
      } else {
        others = true;
      }
    }
    return (match: response, others: others);
  }

  static bool _notEmpty(String? s) => s != null && s.isNotEmpty;

  static String _formatDate(DateTime d) {
    String p2(int v) => v.toString().padLeft(2, '0');
    return '${d.year.toString().padLeft(4, '0')}/${p2(d.month)}/${p2(d.day)} '
        '${p2(d.hour)}:${p2(d.minute)}';
  }

  static DateTime _parseDate(String s) {
    final m = RegExp(
      r'^(\d{4})/(\d{1,2})/(\d{1,2})\s+(\d{1,2}):(\d{1,2})$',
    ).firstMatch(s.trim());
    if (m == null) return DateTime.now();
    return DateTime(
      int.parse(m.group(1)!),
      int.parse(m.group(2)!),
      int.parse(m.group(3)!),
      int.parse(m.group(4)!),
      int.parse(m.group(5)!),
    );
  }
}
